draft-ietf-masque-h3-datagram.md

```diff
@@ -181,74 +181,82 @@ all cases, the maximum permitted value of the H3_DATAGRAM SETTINGS parameter is
+# REGISTER_DATAGRAM_FLOW_ID HTTP/3 Frame Definition {#register-frame}
 
-"Datagram-Flow-Id" is a List Structured
-Field {{!STRUCT-FIELD=I-D.ietf-httpbis-header-structure}}, whose members MUST
-all be Items of type Integer. Integers MUST be non-negative. Its ABNF is:
+The REGISTER_DATAGRAM_FLOW_ID frame (type=TBD) is used to register a flow ID
+with the QUIC stream that it is sent on. By sending this frame, the sender
+indicates that it will process any received datagram with the corresponding
+flow ID using semantics specific to this frame. For example, if this is sent on
+a client-initiated bidirectional stream that has carried an HTTP request, this
+flow ID will now be interpreted as governed by the method of that request.
 
 ~~~
-  Datagram-Flow-Id = sf-list
+REGISTER_DATAGRAM_FLOW_ID Frame {
+  Type (i) = TBD,
+  Length (i),
+  Flow ID (i),
+  Encoded Field Section (..),
+}
 ~~~
+{: #register-frame-format title="REGISTER_DATAGRAM_FLOW_ID HTTP/3 Frame Format"}
 
-The "Datagram-Flow-Id" header field is used to associate one or more datagram
-flows with an HTTP message. As a simple example using a single
-flow, the definition of an HTTP method could instruct the client to use
-its flow ID allocation service to allocate a new flow ID, and
-then the client will add the "Datagram-Flow-Id" header field to its request to
-communicate that value to the server. In this example, the resulting header
-field could look like:
+The Type and Length fields follows the definition of HTTP/3 frames from {{H3}}.
+The payload consists of:
 
-~~~
-  Datagram-Flow-Id = 2
-~~~
+Flow ID:
 
-List members are flow ID elements, which can be named or unnamed.
-One element in the list is allowed to be unnamed, but all but one elements
-MUST carry a name. The name of an element is encoded in the key of the first
-parameter of that element (parameters are defined in Section 3.1.2 of
-{{STRUCT-FIELD}}). Each name MUST NOT appear more than once in the list. The
-value of the first parameter of each named element (whose corresponding key
-conveys the element name) MUST be of type Boolean and equal to true. The value
-of the first parameter of the unnamed element MUST NOT be of type Boolean. The
-ordering of the list does not carry any semantics. For example, an HTTP method
-that wishes to use four datagram flows for the lifetime of its
-request stream could look like this:
+: The flow ID to associate with the QUIC stream that carried this frame.
 
-~~~
-  Datagram-Flow-Id = 42, 44; ecn-ect0, 46; ecn-ect1, 48; ecn-ce
-~~~
+Encoded Field Section:
+
+: QPACK-encoded fields. This allows extensions to optionally attach headers to
+their flow IDs.
+
+Note that these registrations are unilateral and unidirectional: the sender of
+the frame unilateraly defines the semantics it will apply to the datagrams it
+receives. If a mechanism using this feature wants to send datagrams of a given
+flow ID in both directions, this frame will need to be exchanged in both
+directions.
```

Comment:
I understand the spirit of the text but it seems slightly obtuse. It took a few reads and a cross-ref to CONNECT-UDP to help me. Maybe there is a some rewording to be done if we go in this direction.

Reply:
I think the text should also address collisions or re-use of flow IDs.  What happens if the same flow ID is registered to more than one stream?  What is the lifetime of a flow ID -- does it auto-retire at the end of a stream, or can it outlive the FIN?  If the latter, how does the receiver of this frame know that the flow ID is done and state can be removed?